docs/source/coding_conventions.rst

```diff
@@ -0,0 +1,119 @@
+Coding Conventions
+==================
+
+In order to increase readability across Tensor Comprehensions written by
+multiple authors and to reduce the amount of surprising behavior, the
+following conventions should be adopted when writing TC. Generally in TC one
+should increment nesting by 4 whitespaces at each level and align tensor names
+and indices where appropriate to make memory access patterns emerge. Since
+these two goals can easily be conflicting, use your best judgement to tradeoff
+between the two goals. Such examples are provided below.
+
+Use indices named after parameters
+----------------------------------
+
+Use upper-case names for parameters and input/output tensors.
+Use lower-case names for indices to match the name of the parameter
+corresponding to the dimension upon which they iterate.
+In other words, prefer:
+
+.. code::
+
+    def copy2d(float(M, N) I) -> (O) {
+        O(m, n) = I(m, n)
+    }
+
+to:
+
+.. code::
+
+    def copy2d(float(M, N) I) -> (O) {
+        O(i, j) = I(i, j)
+    }
+
+Prefix reduction index names with :code:`r_`
+--------------------------------------------
+
+By definition, reduction indices are the ones that appear on the RHS of a TC
+expression but not on the LHS. On larger expressions it can get challenging to easily
+detect the reduction variables by mentally parsing the set of indices on the
+RHS and subtracting the set of indices on the LHS from it. To alleviate such
+issues, name the reduction variables with a :code:`r_` prefix.
+In other words, prefer:
+
+.. code::
+
+    def matmul(float(M, K) A, float(K, N) B) -> (C) {
+        C(m, n) +=! A(m, r_k) * B(r_k, n)
+    }
+
+to:
+
+.. code::
+
+    def matmul(float(M, K) A, float(K, N) B) -> (C) {
+        C(m, n) +=! A(m, k) * B(k, n)
+    }
+
+Filter non-rectangular regions with deta-dependencies
+-----------------------------------------------------
+
+TC semantics only support (hyper-)rectangular iteration spaces. This is a hard
+requirement to make range inference non-ambiguous. To simulate non-rectangular
+iteration spaces, one can use the following:
+
+.. code::
+
+    def matmul(float(M, K) L, float(K, M) U) -> (LU) {
+        LU(m1, m2) +=! (r_k >= m1 and r_k =< m2) ? L(m1, r_k) * U(r_k, m2) : 0
+    }
+
+However, the following is incompatible with range inference and will fail
```

Comment:
Can we formalize "the following" here?  I don't really like restrictions-by-simple-example.  It seems that "using iteration variables in `where` expression is forbidden" is what you want. 

Also, this is a _semantic_ restriction and not a _style_ recommendation.

Reply:
We can specifically mention that it is a "semantic restriction" and point to the proper place that discusses inference for more details.
Still I think i wouldn't hurt to also surface it in the coding conventions but we can always move it to another section in the docs.